src/plugins/intel_gpu/src/graph/impls/ocl_v2/moe_expert_opt.cpp

```diff
@@ -0,0 +1,1215 @@
+            dnnl_weights[2].oc = _hidden_size;
+            for (int i = 0; i < 3; i++) {
+                if (mlp_params.param[i].scale) {
+                    dnnl_weights[i].scale = convert2dnnl(mlp_params.param[i].scale,
+                                                         {dnnl_weights[i].ic / dnnl_weights[i].ic_group_size, dnnl_weights[i].oc},
+                                                         dnnl::memory::format_tag::ab);
+                }
+                if (mlp_params.param[i].zp) {
+                    dnnl_weights[i].zp = convert2dnnl(mlp_params.param[i].zp,
+                                                      {dnnl_weights[i].ic / dnnl_weights[i].ic_group_size, dnnl_weights[i].oc},
+                                                      dnnl::memory::format_tag::ab);
+                }
+                if (mlp_params.param[i].weight) {
+                    dnnl_weights[i].weight = convert2dnnl(mlp_params.param[i].weight, {dnnl_weights[i].ic, dnnl_weights[i].oc}, dnnl::memory::format_tag::ba);
+                }
+            }
+        }
+
+        auto mask = 1;
+        auto p = std::getenv("CM_MASK");
+        if (p) {
+            mask = std::atoi(p);
+        }
+        // TODO(moe): enable cm support
+        mask = 0;
+        _up_use_cl = !(mask & 1);
+        _down_use_cl = !(mask & 2);
+    }
+
+    void load(BinaryInputBuffer& ib) override {
+        PrimitiveImplOCL::load(ib);
+        const kernel_impl_params* impl_params = reinterpret_cast<kernel_impl_params*>(ib.getKernelImplParams());
+        init(impl_params->typed_desc<moe_expert>());
+    }
+
+    [[nodiscard]] std::unique_ptr<primitive_impl> clone() const override {
+        auto moe = make_deep_copy<MoeExpertOptImpl>(this);
+        moe->_dnnl_weights = _dnnl_weights;
+        moe->_hidden_size = _hidden_size;
+        moe->_intermediate_size = _intermediate_size;
+        moe->_group_size = _group_size;
+        moe->_up_use_cl = _up_use_cl;
+        moe->_down_use_cl = _down_use_cl;
+        return moe;
+    }
+
+    std::vector<BufferDescriptor> get_internal_buffer_descs(const kernel_impl_params& params) const override {
+        auto moe = params.typed_desc<moe_expert>();
+        const auto& config = moe->_config;
+        int max_topk = static_cast<int>(config.topk);
+        int expert_num = static_cast<int>(config.expert_num);
+
+        auto hidden_states_layout = params.input_layouts[0];
+        auto batch = static_cast<int>(hidden_states_layout.get_shape()[0]);
+        auto data_type = hidden_states_layout.data_type;
+
+        std::vector<BufferDescriptor> internal_buffers;
+        // softmax+topk
+        layout layout_topk_id(ov::PartialShape{batch, max_topk}, data_types::u32, cldnn::format::bfyx);
+        layout layout_topk_weights(ov::PartialShape{batch, max_topk}, data_type, cldnn::format::bfyx);
+        internal_buffers.emplace_back(layout_topk_id, false);       // topk_id
```

Comment:
Considering that we need to access the memory from the host, it might be better to use usm_host here

Reply:
update